### PR TITLE
Fix Issue 7879 - format of const class with non const toString()

### PR DESCRIPTION
**Pull request by JackStouffer**
So trying to make `Object.toString` const was a nightmare, so I added in this work-around to at least allow the code to compile.

### CODE REVIEW

**Comment by JackStouffer** [comment]
Changed code to match std.format's existing behavior for class formatting

**Comment by wilzbach** [comment]
> So trying to make Object.toString const was a nightmare

Yeah, though we should really move things out of `object.d`.
A quick selection:

- https://issues.dlang.org/show_bug.cgi?id=9769
- https://issues.dlang.org/show_bug.cgi?id=9770
- https://issues.dlang.org/show_bug.cgi?id=9771
- https://issues.dlang.org/show_bug.cgi?id=9772

File: std/format.d [hunk 1]
```diff
@@ -3830,6 +3841,53 @@ if (is(T == class) && !is(T == enum))
...
+    s = format("%s", c4);
+    assert(s == "shared(std.format.C)");
+}
+
+// https://issues.dlang.org/show_bug.cgi?id=7879
+@safe unittest
+{
+    class F
+    {
+        override string toString() const @safe
```
Comment (wilzbach) [inline std/format.d]: BTW we can't use `inout` here: https://issues.dlang.org/show_bug.cgi?id=18500

File: std/format.d [hunk 2]
```diff
@@ -3738,7 +3736,20 @@ if (is(T == class) && !is(T == enum))
         put(w, "null");
     else
     {
-        static if (hasToString!(T, Char) > 1 || (!isInputRange!T && !is(BuiltinTypeOf!T)))
+        static if ((is(T == immutable) || is(T == const) || is(T == shared)) && hasToString!(T, Char) == 0)
```
Comment (wilzbach) [inline std/format.d]: I would prefer if we could use `== none` here.
See: https://github.com/dlang/phobos/pull/6209

File: std/format.d [hunk 3]
```diff
@@ -3738,7 +3736,20 @@ if (is(T == class) && !is(T == enum))
...
+                put(w, "immutable(");
+            else static if (is(T == const))
+                put(w, "const(");
+            else static if (is(T == shared))
+                put(w, "shared(");
+
+            put(w, typeid(Unqual!T).name);
+            put(w, ')');
+        }
+        else static if (hasToString!(T, Char) > 1 || !isInputRange!T && !is(BuiltinTypeOf!T))
```
Comment (wilzbach) [inline std/format.d]: With https://github.com/dlang/phobos/pull/6209: `!overload.among(none, hasSomeToString)`

**Comment by JackStouffer** [comment]
Moving ahead with this to shorten the queue.

**Comment by JackStouffer** [comment]
@wilzbach Can I get a manual merge?

**Pull request merged**